relay-quotas/src/global.rs

```diff
@@ -0,0 +1,256 @@
+                // TODO double check logic here
+                // in theory this should never happen only if someone messes with the system time
+                // be safe and dont rate limit
+                relay_log::error!("budget slot ahead of current slot");
+                Ok(false)
+            }
+            Ordering::Less => {
+                self.counter = Counter::new();
+                self.slot = current_slot;
+
+                self.counter.is_rate_limited(client, limit, quantity, quota)
+            }
+            Ordering::Equal => self.counter.is_rate_limited(client, limit, quantity, quota),
+        }
+    }
+}
+
+impl Default for SlottedCounter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Represents the local budget taken from a global quota.
+struct Counter {
+    local_counter: LocalCounter,
+    redis_counter: RedisCounter,
+}
+
+impl Counter {
+    pub fn new() -> Self {
+        Self {
+            local_counter: LocalCounter::new(),
+            redis_counter: RedisCounter::new(),
+        }
+    }
+
+    pub fn is_rate_limited(
+        &mut self,
+        client: &mut PooledClient,
+        limit: u64,
+        quantity: usize,
+        quota: &RedisQuota,
+    ) -> Result<bool, RedisError> {
+        if self.local_counter.try_consume(quantity) {
+            return Ok(false);
+        }
+
+        if !self.redis_counter.can_satisfy(quantity, limit) {
+            return Ok(true);
+        }
+
+        let budget_to_reserve = quantity.max(self.default_request_size_based_on_limit());
+        let reserved = self
+            .redis_counter
+            .try_reserve(client, budget_to_reserve, limit, quota)? as usize;
+
+        self.local_counter.increase_budget(reserved);
+        Ok(!self.local_counter.try_consume(quantity))
+    }
+
+    fn default_request_size_based_on_limit(&self) -> usize {
+        100
+    }
+}
+
+struct LocalCounter {
+    budget: usize,
+}
+
+impl LocalCounter {
+    fn new() -> Self {
+        Self { budget: 0 }
+    }
+
+    fn try_consume(&mut self, quantity: usize) -> bool {
+        if self.budget >= quantity {
+            self.budget -= quantity;
+            true
+        } else {
+            false
+        }
+    }
+
+    fn increase_budget(&mut self, quantity: usize) {
+        self.budget = self.budget.saturating_add(quantity);
+    }
+}
+
+struct RedisCounter {
+    last_seen: u64,
+}
+
+impl RedisCounter {
+    fn new() -> Self {
+        Self { last_seen: 0 }
+    }
+
+    fn can_satisfy(&self, quantity: usize, limit: u64) -> bool {
+        self.last_seen + quantity as u64 <= limit
+    }
+
+    fn try_reserve(
+        &mut self,
+        client: &mut PooledClient,
+        quantity: usize,
+        limit: u64,
+        quota: &RedisQuota,
+    ) -> Result<u64, RedisError> {
+        let script = load_global_lua_script();
+
+        let redis_key = quota.key();
+        let expiry = UnixTimestamp::now().as_secs() + quota.window();
```

Comment:
`quota.expiry()` exists already, is there a reason we cannot use that?

In `redis.rs` we use `GRACE`, we should also do it here. To not have it in multiple spots, we can just add the `GRACE` to the return value of `quota.expiry()` directly.